apps/api/src/app/subscribers/usecases/handle-chat-oauth/handle-chat-oauth.usecase.ts

```diff
@@ -0,0 +1,126 @@
+  readonly SLACK_ACCESS_URL = 'https://slack.com/api/oauth.v2.access';
+  readonly SCRIPT_CLOSE_TAB = '<script>window.close();</script>';
+
+  constructor(
+    private updateSubscriberChannelUsecase: UpdateSubscriberChannel,
+    private integrationRepository: IntegrationRepository,
+    private environmentRepository: EnvironmentRepository,
+    private createSubscriberUsecase: CreateSubscriber
+  ) {}
+
+  async execute(command: HandleChatOauthCommand) {
+    const { credentials: integrationCredentials } = await this.getIntegration(command);
+
+    const webhookUrl = await this.getWebhook(command, integrationCredentials);
+
+    const environment = await this.getEnvironment(command);
+
+    await this.createSubscriber(environment, command, webhookUrl);
+
+    const redirect = integrationCredentials.redirectUrl != null && integrationCredentials.redirectUrl != '';
+
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    return { redirect, action: redirect ? integrationCredentials.redirectUrl! : this.SCRIPT_CLOSE_TAB };
+  }
+
+  private async createSubscriber(environment, command: HandleChatOauthCommand, webhookUrl) {
+    await this.createSubscriberUsecase.execute(
+      CreateSubscriberCommand.create({
+        organizationId: environment._organizationId,
+        environmentId: command.environmentId,
+        subscriberId: command?.subscriberId,
+      })
+    );
+
+    const subscriberCredentials: IChannelCredentialsCommand = { webhookUrl: webhookUrl, channel: command.providerId };
+
+    await this.updateSubscriberChannelUsecase.execute(
+      UpdateSubscriberChannelCommand.create({
+        organizationId: environment._organizationId,
+        environmentId: command.environmentId,
+        subscriberId: command.subscriberId,
+        providerId: command.providerId,
+        credentials: subscriberCredentials,
+        oauthHandler: OAuthHandlerEnum.NOVU,
+      })
+    );
+  }
+
+  private async getEnvironment(command: HandleChatOauthCommand) {
+    const environment = await this.environmentRepository.findById(command.environmentId);
+
+    if (environment == null) {
+      throw new ApiException(`Environment ID: ${command.environmentId} not found`);
+    }
+
+    return environment;
+  }
+
+  private async getWebhook(command: HandleChatOauthCommand, integrationCredentials: ICredentialsDto) {
+    const redirectUri =
+      process.env.API_ROOT_URL +
+      `/v1/subscribers/${command.subscriberId}/${command.providerId}/${command.environmentId}`;
+
+    const body = {
+      redirect_uri: redirectUri,
+      code: command.providerCode,
+      client_id: integrationCredentials.clientId,
+      client_secret: integrationCredentials.secretKey,
+    };
+    const config = {
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+    };
+
+    const res = await axios.post(this.SLACK_ACCESS_URL, body, config);
+    const webhook = res.data?.incoming_webhook?.url;
```

Comment:
I consider this should be an application generic Slack service. But so far could it live here by now.

Reply:
Yeah once we will have this logic in multiple providers we could refactor it.